src/common/servers/serverManager.ts

```diff
@@ -0,0 +1,426 @@
+    }
+
+    getLastActiveTabForServer = (serverId: string) => {
+        log.debug('ServerManager.getLastActiveTabForServer', serverId);
+
+        const lastActiveTab = this.lastActiveTab.get(serverId);
+        if (lastActiveTab) {
+            const tab = this.tabs.get(lastActiveTab);
+            if (tab && tab?.isOpen) {
+                return tab;
+            }
+        }
+        return this.getFirstOpenTabForServer(serverId);
+    }
+
+    getServer = (id: string) => {
+        return this.servers.get(id);
+    }
+
+    getTab = (id: string) => {
+        return this.tabs.get(id);
+    }
+
+    getAllServers = () => {
+        return [...this.servers.values()];
+    }
+
+    hasServers = () => {
+        return Boolean(this.servers.size);
+    }
+
+    getRemoteInfo = (serverId: string) => {
+        return this.remoteInfo.get(serverId);
+    }
+
+    updateRemoteInfos = (remoteInfos: Map<string, RemoteInfo>) => {
+        let hasUpdates = false;
+        remoteInfos.forEach((remoteInfo, serverId) => {
+            this.remoteInfo.set(serverId, remoteInfo);
+            hasUpdates = this.updateServerURL(serverId) || hasUpdates;
+            hasUpdates = this.openExtraTabs(serverId) || hasUpdates;
+        });
+
+        if (hasUpdates) {
+            this.persistServers();
+        }
+    }
+
+    lookupTabByURL = (inputURL: URL | string, ignoreScheme = false) => {
+        log.silly('lookupTabByURL', `${inputURL}`, ignoreScheme);
+
+        const parsedURL = urlUtils.parseURL(inputURL);
+        if (!parsedURL) {
+            return undefined;
+        }
+        const server = this.getAllServers().find((server) => {
+            return equalUrlsIgnoringSubpath(parsedURL, server.url, ignoreScheme) && parsedURL.pathname.match(new RegExp(`^${server.url.pathname}(.+)?(/(.+))?$`));
+        });
+        if (!server) {
+            return undefined;
+        }
+        const tabs = this.getOrderedTabsForServer(server.id);
+
+        let selectedTab = tabs.find((tab) => tab && tab.name === TAB_MESSAGING);
+        tabs.
+            filter((tab) => tab && tab.name !== TAB_MESSAGING).
+            forEach((tab) => {
+                if (parsedURL.pathname.match(new RegExp(`^${tab.url.pathname}(/(.+))?`))) {
+                    selectedTab = tab;
+                }
+            });
+        return selectedTab;
+    }
+
+    updateServerOrder = (serverOrder: string[]) => {
+        log.debug('updateServerOrder', serverOrder);
+
+        this.serverOrder = serverOrder;
+        this.persistServers();
+    }
+
+    updateTabOrder = (serverId: string, tabOrder: string[]) => {
+        log.debug('ServerManager.updateTabOrder', serverId, tabOrder);
+
+        this.tabOrder.set(serverId, tabOrder);
+        this.persistServers();
+    }
+
+    addServer = (server: Team) => {
+        const newServer = new MattermostServer(server, false);
+        this.servers.set(newServer.id, newServer);
+        this.serverOrder.push(newServer.id);
```

Comment:
if the identifier on the map already exists, you may be replacing an existing server. but will have the same id as duplicate in the `serverOrder` array. can we be a little bit safer here?

Reply:
If we ever get the same ID somehow, that would mean that our guaranteed UUIDs have collided. It's really unlikely and should cause errors to happen anyways, but I can add an explicit error check here if we think it's necessary.